src/cpp/fastdds/publisher/DataWriterImpl.cpp

```diff
@@ -1218,10 +1215,16 @@ ReturnCode_t DataWriterImpl::set_qos(
         return RETCODE_IMMUTABLE_POLICY;
     }
 
+    // Take a snapshot of the current QoS before mutating it
+    const DataWriterQos old_qos = qos_;
+
     set_qos(qos_, qos_to_set, !enabled);
 
     if (enabled)
     {
+        std::unique_lock<RecursiveTimedMutex> lock(writer_->getMutex());
+        // Locks after we checked that writer exists
```

Comment:
Don't take the mutex here, it is not thread safe as it is but let's tackle that task in other PR with the rest of changes needed.